lib/pbench/test/unit/server/test_cache_manager.py

```diff
@@ -1552,6 +1553,152 @@ def read(self, size: int = -1) -> bytes:
         assert stream.lock is None
         assert my_calls == exp_calls
 
+    def test_get_contents(self, server_config, make_logger, tarball, db_session):
+        """Test the TOC analysis"""
+
+        source_tarball, source_md5, md5 = tarball
+        cm = CacheManager(server_config, make_logger)
+        cm.create(source_tarball)
+        root = f"https://fake/api/v1/datasets/{md5}"
+
+        # Unpack the tarball and show the canonical `tarball` fixture contents
+        assert cm.get_contents(md5, "", root) == {
+            "name": "",
+            "type": "DIRECTORY",
+            "uri": f"{root}/contents/",
+            "directories": [],
+            "files": [
+                {
+                    "name": "metadata.log",
+                    "size": 28,
+                    "type": "FILE",
+                    "uri": f"{root}/inventory/metadata.log",
+                }
+            ],
+        }
+
+        assert cm.get_contents(md5, "metadata.log", root) == {
+            "name": "metadata.log",
+            "type": "FILE",
+            "uri": f"{root}/inventory/metadata.log",
+            "size": 28,
+        }
+
+        with pytest.raises(CacheExtractError):
+            cm.get_contents(md5, "../..", root)
+
+        # Now that we have an unpacked directory, manually "enhance" it to test
+        # more paths.
+
+        base: Path = cm.find_dataset(md5).unpacked
+        subdir = base / "subdir"
+        subdir.mkdir()
+        ldir = base / "dir_link"
+        ldir.symlink_to(subdir)
+        lfile = base / "file_link"
+        lfile.symlink_to(Path(base / "metadata.log"))
+        badlink = base / "bad_link"
+        badlink.symlink_to("/etc/passwd")
```

Comment:
This is a symlink with an absolute path; would it make sense to have a bad symlink with a relative path (i.e., an attempt to "walk up" out of the tarball)?

Reply:
I did at one point. I converted to an absolute path because `Path.absolute()` doesn't drop a relative `..` and `relative_to` just drops everything up to the `..`, meaning we're silently going outside the boundaries. I figured out a solution by going to `os.path.abspath`, which *does* drop the `..` so that `relative_to` will fail ... but apparently forgot to reinstate the test case. 😦